crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -342,6 +345,14 @@ pub unsafe trait QueryData: WorldQuery {
         entity: Entity,
         table_row: TableRow,
     ) -> Option<Self::Item<'w, 's>>;
+
+    /// Returns an iterator over the access needed by [`QueryData::fetch`]. Access conflicts are usually
+    /// checked in [`WorldQuery::init_fetch`], but in certain cases this method can be useful to implement
+    /// a way of checking for access conflicts in a non-allocating way.
+    fn iter_access<'a>(
```

Comment:
Would it make sense to have the caller do `get_state` first and pass in the `&Self::State` here?  

They're almost certainly about to call it anyway, and that would let us avoid needing to do the `HashMap` lookup in `components.component_id::<T>()` twice.  Or even more than twice, since `has_conflicts` calls this in a loop.  

And that would let you handle `FilteredEntityMut` and `EntityMutExcept` by including a variant of `EcsAccessType` that borrows the `&Access`.  

It would be a little awkward to hook in to `get_components_mut` since `get_state` is called inside `UnsafeEntityCell::get_components`.  Although even if you wind up calling `get_state` twice, that's no more calls to `component_id::<T>()` than you're doing now.

Reply:
I have no idea what that would look like. How would you convert the get_state to an iterator?

Reply:
You convinced me. Note that since Entity*Except and FilteredEntity* do allocate, using them in a uncached query might be a perf footgun. The issue with allocating once a frame isn't just the normal cost of allocation, but potentially fragmenting your memory which can cause allocation costs to go up significantly.